[Tan Yu Tao] ip

tanyutao544 opened the pull request:
# **Duke V2.0**
> *Now with more tape*

## Duke V2.0 is Duke but with more features and tape to hold it together.
Duke V2.0 is:
- Text-based
- Easy to learn, ~~tedious to use~~
- Fast to setup

All you need to do is,

1. download it from [here](https://github.com/tanyutao544/ip/releases/tag/A-Jar)
2. double-click it
3. add your tasks
4. let it manage your tasks for you :sunglasses:

And it is **FREE!**

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

It is completely open-source too, here's a code snippet
```
public static void main(String[] args) {
		new Duke("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

tanyutao544 committed:
Setup skeletal version of Duke

tanyutao544 committed:
Added basic Commands

tanyutao544 committed:
Modified to accomodate new Commands

tanyutao544 committed:
Added new Task class

tanyutao544 committed:
Updated commands to accomodate new Task class

tanyutao544 committed:
Added new Mark and Unmark command class

tanyutao544 committed:
Implemented new Task class and new commands

tanyutao544 committed:
Modified addcommand to use Task toString method

tanyutao544 committed:
Added Deadline, Events, Todos

tanyutao544 committed:
Modified to inherit from Task and print appropriate detail

tanyutao544 committed:
Resolve formatting

tanyutao544 committed:
Modified Task class to be abstract

tanyutao544 committed:
Renamed method variables to be clearer

tanyutao544 committed:
Modified Addcommand to create appropriate Task based on input

tanyutao544 committed:
Added input and expected output for automated testing

tanyutao544 committed:
Added new DukeException abstract class and EmptyDescription, UnkownCommand Excpetion class

tanyutao544 committed:
Modified Addcommand to store type of Task to add

tanyutao544 committed:
Modified duke processinput to throw new exceptions, and main to catch and display exception error messages

tanyutao544 committed:
Renamed DukeExceptions class to DukeException and resolved all relevant references

tanyutao544 committed:
Cleaned up code for changes in class name and streamlined main program body

tanyutao544 committed:
Added skeletal structure of deleteCommand

tanyutao544 committed:
Completed DeleteCommand

tanyutao544 committed:
Implemented Delete command into Duke main

tanyutao544 committed:
Modified methods to accept commands regardless of capitalization

tanyutao544 committed:
Modified to use Arraylist

tanyutao544 committed:
Updated all commands to use Arraylist

tanyutao544 committed:
Added 2 new exceptions skeletons for mark, unmark and delete command

tanyutao544 committed:
Completed IncorrectValueException and OutOfRangeException

tanyutao544 committed:
Fixed bug of referencing wrong deleteIndex when getting deleted task

tanyutao544 committed:
Implemented new exceptions to be thrown when checking input for mark, unmark and delete command

tanyutao544 committed:
Update A-TextUiTesting input.txt and EXPECTED.txt

Current input and expected output text do not test behaviour of new commands and error catching

Update input txt and expected txt to test behaviour when inputing wrong command, without description/wrong description and capitalization of commands.

tanyutao544 committed:
Update coding style

Switch case indentation style, case is indented from switch statement

Update style where case is on same indentation level as switch statement

tanyutao544 committed:
Update processInput to throw parent exception

Method threw subclass exceptions when parent exception could be thrown to simplify code for reading

Method throws parent exception inherited by child exceptions

tanyutao544 committed:
Update Tasks naming convention, add getter functions to respective class type

Todos Events Deadlines class name in plural no getter function for respective class type

Update class name to singular add getter function for detail, date, mark for respective class types

tanyutao544 committed:
Update AddCommand to reflect changes to class naming convention changes, Add getter function to parent class Task

No getter function for task class, AddCommand task referncing in plural

Update AddCommand to reference correct class name, Add getter function for Task Class

tanyutao544 committed:
Update gitignore to ignore duke.txt file

Did not ignore duke.txt, pushes changes to duke,txt to repo

Now ignores duke.txt when commiting and pushing to repo

tanyutao544 committed:
Add savefile and loadfile functionality

Did not save list generated when running program

Add functionality to generate file if not in local folder and saves and loads from the file at start and end of program run

tanyutao544 committed:
Update Deadlines, Events class to read date as LocalDateTime format

Deadlines, Events stored date as String

Update date to LocalDateTime format and edited constructors to store as correct format

tanyutao544 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/tasks/Deadline.java
#	src/main/java/tasks/Event.java

tanyutao544 committed:
Resolve error from getDate method

Method return String when deadline/date stored as LocalDateTime

Update Method to return string in proper format for saving and loading Duke.txt

tanyutao544 committed:
Add new Storage class

Saving and loading files use methods in duke.java

Abstract methods from duke.java into new Storage class

tanyutao544 committed:
Add Ui class

Duke main prints welcome logo

Abstract logo and welcome command into Ui class specific method

tanyutao544 committed:
Update tasks to print without lines abstract to Ui class

Ui class did not handle line printing

Ui class prints the line for recursive use in run method

tanyutao544 committed:
Refactor Ui and Storage to individual package

Classes in same folder as Duke.java

Package into individual package folder

tanyutao544 committed:
Refactor package names and add Tasklist class encompassing add, remove functions of a task list

Naming convention on packages not consistent, Duke uses Arraylist to store tasks

Add dedicated class for storing class and standardize naming of packages

tanyutao544 committed:
Update Command classes to accpet inputs during execute method instead of during construction of class.

Constructor takes in tasks not coinciding with execute method naming

Update methods and variables in each command class to make use of new Ui. bye command now stores the tasks into task.txt during execution

tanyutao544 committed:
update DukeException error message printing to coincide with Ui printing. update Tasklist to have set method

Tasklist did not have method to replace items inside the list. Some ui bugs with line printing

Remedy Ui printing bugs and adds method to replace item in tasklist

tanyutao544 committed:
Factors out processInput method in Duke to its own parser class modify duke to make use of new parser class

Duke main used to have own process input method, cluttering the code

Abstracts methods related to processing input into its own Parser class

tanyutao544 committed:
Update Exceptions to override getMessage from execeptions instead of abstract message in dukeException class. Resolved bug when deleting, nullpointer exception, set TaskList variable to non static

Error when attempting to delete a valid index, Exception classes were printing messages instead of the convention of returning a string

DeleteCommand was not copying over tasks parsed in execute method. Standardize errors to return string instead of printing the error message.

tanyutao544 committed:
Add testcases for commands

No unit testing for commands implemented

Adds unit testing for all current commands

tanyutao544 committed:
Add Unit testing for Exceptions

No unit testing for exceptions

Add unit testing for exceptions to check if correct error message is printed

tanyutao544 committed:
Add Java doc to all classes

JavaDoc missing from classes

Adds javadoc to every class and method

tanyutao544 committed:
Add FindCommand class to find tasks using a keyword

No method to find tasks in a list using keyword

Adds command that finds a task in the list using a keyword

tanyutao544 committed:
Standardizes coding standard

Some coding standard violations

Rectifies some violations

tanyutao544 committed:
Merge tag 'Level-9'

tanyutao544 committed:
Merge tag 'A-JavaDoc'

tanyutao544 committed:
Update command method ends to endsProgram to sound more boolean

Method ends in command class does not sound boolean enough

Update test cases and command class method to endsProgram

tanyutao544 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

tanyutao544 committed:
Add gradle support for building and testing

No support for gradle to run unit tests

tanyutao544 committed:
Fix coding style

tanyutao544 committed:
Merge branch 'branch-Gradle' add support got checkstyle in gradle

tanyutao544 committed:
Fix coding style in test cases

tanyutao544 committed:
Resolve style check violations

tanyutao544 committed:
Resolve coding style violations for Test unit

tanyutao544 committed:
Adds javadoc to Junit Test classes

tanyutao544 committed:
Adds JavaFX support

tanyutao544 committed:
Resolve coding style violations in GUI feature

sbhbenjamin commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,81 @@
+    @Override
+    public boolean endsProgram() {
+        return false;
+    }
```
Instead of adding this piece of code to check if the code is an exit type command, would it be better to give each command a name field, and Duke can check if the returned command's name is of the exit command? I thought that this might help in extensibility, such that the name field can be used for other purposes in the future (i.e. checking what type of command it is)

I thought this was a pretty good way to do it!: https://github.com/se-edu/addressbook-level2/blob/master/src/seedu/addressbook/commands/AddCommand.java

sbhbenjamin commented on src/main/java/duke/exceptions/OutOfRangeException.java:
```diff
@@ -0,0 +1,19 @@
+     * Error messsge from exception
+     * @return String containing details of error
+     */
+    @Override
+    public String getMessage() {
+        return "OOPS!!! The value input is not in the list";
+    }
```
Should you consider passing the string into the parent's constructor using `super` instead of having to override the `getMessage()` function?

sbhbenjamin commented on src/main/java/duke/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,69 @@
+        return "Noted. I've removed this task:\n"
+            + "  " + deleted + "\n"
+            + "Now you have " + this.tasks.size() + " tasks in the list";
+    }
```
I really like how you used encapsulation here to contain all of the logic into the `execute` command. (i.e. updating the storage and returning the string to be printed by the ui) I think this illustrates strong understanding of OOP principles! 👍

sbhbenjamin commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,56 @@
+ * Class which handles all printing and formatting for messages
+ */
+public class Ui {
+    public static final String LINE = "    ____________________________________________________________\n";
```
Its a small thing but I totally forgot this was possible, great use of the DRY principle so that you don't have to keep repeating the line throughout the code!

sbhbenjamin commented on src/main/java/duke/exceptions/DukeException.java:
```diff
@@ -0,0 +1,8 @@
+/**
+ * abstract class from which all exception inherit from
+ */
+public abstract class DukeException extends Exception {
+
+}
```
Should you put the curly braces on the same line instead? i.e. line 6

sbhbenjamin reviewed (commented):
Overall, I found the code to be very well-written and exhibited strong OOP principles. I also really liked the depth of testing conducted, and other than some suggestions on minor matters, LGTM 👍

tanyutao544 committed:
Refactor project to new javafx GUI

There are several methods in duke that are designed for CLI
such methods are removed

Class Ui handled printing of messages from commands, class now
refactored to contain all messages from command

Command classes method refactor to reflect changes in Ui class.
Execution of command also now saves tasklist. Saving changes
in situation where program crashes.

By implementing javafx, code can be further simplified and
refactored. Doing so will allow future changes and features
to be easily implemented

tanyutao544 committed:
Add assertion to commands

tanyutao544 committed:
Merge pull request #1 from tanyutao544/branch-A-Assertion

Add assertion to commands

tanyutao544 committed:
Code quality tidied up

tanyutao544 committed:
Merge pull request #2 from tanyutao544/branch-A-CodeQuality

Code quality tidied up

tanyutao544 committed:
Add new update feature

Allow user to update details or date of a task
by typing update <index>/detail/time/<Update detail>

tanyutao544 committed:
Fixes code quality issues in headers and boolean variable naming

tanyutao544 committed:
update GUI to have dialog container not compress

tanyutao544 committed:
Refactors GUI dialog container change to fxml file

tanyutao544 committed:
Update MainWindow fxml to allow resizing of GUI

tanyutao544 committed:
Update GUI with background and text box

tanyutao544 committed:
Adds user guide and fix bug with deadline type tasks printing
wrong type of task

tanyutao544 committed:
Add Ui screenshot for submission

tanyutao544 committed:
Remove old readme file

tanyutao544 committed:
Renamed Ui.png to resolve issue with product site

tanyutao544 committed:
Fix bug for update command

update user guide with use case for updating time

tanyutao544 committed:
Add new wrong  dateformat exception

Update  relevant commands to throw the new exception
abstract LocalDateTimeformatting out of Task classes

tanyutao544 committed:
Bug fix for addcommand to catch incorrect input formats

tanyutao544 committed:
Apply SLAP to addcommand class
Standardize javadocs for all classes

tanyutao544 committed:
Update test case for addcommand to catch dukeexception

tanyutao544 committed:
fix test cases
